lib/src/revset.rs

```diff
@@ -4292,6 +4471,131 @@ mod tests {
         "#);
     }
 
+    #[test]
+    fn test_optimize_heads_range() {
+        let settings = insta_settings();
+        let _guard = settings.bind_to_scope();
+
+        // Heads of basic range operators can be folded
+        insta::assert_debug_snapshot!(optimize(parse("heads(::)").unwrap()), @r"
+        HeadsRange {
+            roots: None,
+            heads: VisibleHeads,
+            predicates: [],
+        }
+        ");
+        insta::assert_debug_snapshot!(optimize(parse("heads(::foo)").unwrap()), @r#"
```

Comment:
nit: can you add test for `heads(::foo-)`? `::foo-` will be transformed to `ancestors(foo, depth=1..)`.